source/merge-deep.d.ts

```diff
@@ -0,0 +1,136 @@
+interface MergeDeepOptions {
+  strict?: boolean;
+}
+
+type UnknownRecord = Record<string, unknown>;
+
+type Unwrap<Type> = Type extends UnknownRecord
+  ? {[Key in keyof Type]: Type[Key]}
+  : Type;
+
+type MergeValue<Destination, Source, Key, Options> =
+  Key extends keyof Destination
+    ? Key extends keyof Source
+      ? Destination[Key] extends UnknownRecord
+        ? MergeDeep<Destination[Key], Source[Key], Options>
+        : Source[Key]
+      : Destination[Key]
+    : Key extends keyof Source
+      ? Source[Key]
+      : never;
+
+type Keyof<Destination, Source> = keyof Destination | keyof Source;
+
+type MergeArrayValue<Destination, Source, Options> =
+  Destination extends UnknownArray
+    ? Source extends UnknownArray
+      ? ArrayMerge<Destination, Source, Options>
+      : Source
+    : Source;
+
+type GetValue<
+  Destination,
+  Source,
+  Key extends Keyof<Destination, Source>,
+  Options,
+> = Key extends keyof Source
+  ? Source[Key] extends UnknownRecord
+    ? MergeValue<Destination, Source, Key, Options>
+    : Source[Key] extends UnknownArray
+			? Key extends keyof Destination
+				? MergeArrayValue<Destination[Key], Source[Key], Options>
+				: Source[Key]
+			: Source[Key]
+  : Key extends keyof Destination
+  ? Destination[Key]
+  : never;
+
+type ExtractType<TType, TUnion> = {
+  [Key in keyof TType]-?: TUnion extends TType[Key] ? Key : never;
+}[keyof TType];
+
+type OmitType<TType, TUnion> = Omit<TType, ExtractType<TType, TUnion>>;
+
+type MergeDeepLazy<Destination, Source, Options> = {
+  [Key in Keyof<Destination, Source>]: Unwrap<
+    GetValue<Destination, Source, Key, Options>
+  >;
+};
+
+type MergeDeepStrict<Destination, Source, Options> = OmitType<
+  MergeDeepLazy<Destination, Source, Options>,
+  undefined
+>;
+
+type ArrayHead<TArray> = TArray extends readonly [infer THead, ...unknown[]]
```

Comment:
Move it to https://github.com/sindresorhus/type-fest/blob/main/source/internal.d.ts and add doc comment.

Reply:
For the moment I have moved `ArrayHead` and `ArrayTail` to internal. But as the `LastArrayElement` type exists, I could add them as complementary types.

- FirstArrayElement (aka ArrayHead)
- RestArrayElements (aka ArrayTail)

And maybe later (if it's possible, I haven't thought about it):
- NFirstArrayElements
- NLastArrayElements